README.md
# @openmev/ethers-provider

> Multi provider support enabled ethers web3 provider

## Overview

- [see documentation for differences and additional functionality](https://manifoldfinance.github.io/ethers-provider)


## License

MIT


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v0.4.2](https://github.com/manifoldfinance/ethers-provider/compare/v0.4.0...v0.4.2)

> 23 September 2021

- Lint [`#50`](https://github.com/manifoldfinance/ethers-provider/pull/50)
- fix minimumNonceByAccount [`#40`](https://github.com/manifoldfinance/ethers-provider/pull/40)
- 0.4.1 [`#46`](https://github.com/manifoldfinance/ethers-provider/pull/46)
- Add getBundleStats, use in demo.ts, and update docs [`#44`](https://github.com/manifoldfinance/ethers-provider/pull/44)

#### [v0.4.0](https://github.com/manifoldfinance/ethers-provider/compare/v0.3.1...v0.4.0)

> 27 July 2021

- Update deps [`#43`](https://github.com/manifoldfinance/ethers-provider/pull/43)
- Bump ethers to 5.4.2, slight customization to bundle submission to support transaction type 2 [`#42`](https://github.com/manifoldfinance/ethers-provider/pull/42)
- Fix stateBlockNumber simulation parameter [`#37`](https://github.com/manifoldfinance/ethers-provider/pull/37)
- Add support for `error` field being `null` [`#34`](https://github.com/manifoldfinance/ethers-provider/pull/34)
- Add to README.md and add github repo url to npm package [`#33`](https://github.com/manifoldfinance/ethers-provider/pull/33)
- Updates for security [`59ffe37`](https://github.com/manifoldfinance/ethers-provider/commit/59ffe37047212fbd49c8baa7350a8d47a91b141a)
- Add EIP1559 transaction type=2 demo, add static function to calculate max baseFee in a future block [`f555049`](https://github.com/manifoldfinance/ethers-provider/commit/f55504988807bcbbccf744ef02c2088b62fd6263)
- Try fixing lint [`414a7a1`](https://github.com/manifoldfinance/ethers-provider/commit/414a7a18018fc63d56fff5dd80200688ea6a2b5d)

#### [v0.3.1](https://github.com/manifoldfinance/ethers-provider/compare/v0.3.0...v0.3.1)

> 10 May 2021

- Searching through all block's Tx instead of lasts [`#32`](https://github.com/manifoldfinance/ethers-provider/pull/32)
- lint [`309be15`](https://github.com/manifoldfinance/ethers-provider/commit/309be15ed7b468e823aceef45b67e483f82d5c67)

#### [v0.3.0](https://github.com/manifoldfinance/ethers-provider/compare/v0.2.0...v0.3.0)

> 7 May 2021

- goerli README section [`#31`](https://github.com/manifoldfinance/ethers-provider/pull/31)
- Update various dev dependencies [`#30`](https://github.com/manifoldfinance/ethers-provider/pull/30)
- Update README and provide example of revertingTxHashes [`#28`](https://github.com/manifoldfinance/ethers-provider/pull/28)
- Update ethers to 5.1.4 . Support for EIP-2718 / EIP-2930 [`#29`](https://github.com/manifoldfinance/ethers-provider/pull/29)
- v0.2 Updates [`#27`](https://github.com/manifoldfinance/ethers-provider/pull/27)
- add missing type [`#26`](https://github.com/manifoldfinance/ethers-provider/pull/26)
- update readme [`#25`](https://github.com/manifoldfinance/ethers-provider/pull/25)
- Add new getUserStats method [`#24`](https://github.com/manifoldfinance/ethers-provider/pull/24)
- Add type for getUserStats, minTimestamp for simulate [`e4ee480`](https://github.com/manifoldfinance/ethers-provider/commit/e4ee480fc8f5028c31e11fb94a762fe9263d0cc5)
- Remove round trips in simulate [`e96963b`](https://github.com/manifoldfinance/ethers-provider/commit/e96963b6b4cb9d3ad76da6ebe775d90f7efbd552)
- Update eth_sendBundle for v0.2 [`4104971`](https://github.com/manifoldfinance/ethers-provider/commit/410497111fb43f388563218170c273cd951d483c)

#### [v0.2.0](https://github.com/manifoldfinance/ethers-provider/compare/v0.1.2...v0.2.0)

> 8 March 2021

- Better support for protocol-level errors from callBundle [`#23`](https://github.com/manifoldfinance/ethers-provider/pull/23)
- Export all the types [`#22`](https://github.com/manifoldfinance/ethers-provider/pull/22)
- replace apikeys with signature [`#18`](https://github.com/manifoldfinance/ethers-provider/pull/18)
- 0.1.2 [`#20`](https://github.com/manifoldfinance/ethers-provider/pull/20)
- Fix request formation [`0851e3a`](https://github.com/manifoldfinance/ethers-provider/commit/0851e3a8707a1fe3f06b268a3107fcb106e2f2c6)

test/index.ts

"use strict";

import "./test.providers";

import { LogFunc, Reporter, setLogFunc } from "./reporter";

export { LogFunc, Reporter, setLogFunc }


test/reporter.ts
/* istanbul ignore file */
'use strict';


// Maximum time in seconds to suppress output
const MAX_DELAY = 30;


function getTime(): number {
    return (new Date()).getTime();
}


function trunc(value: number): number {
    if (value >= 0) { return Math.floor(value); }
    return Math.ceil(value);
}


function getDelta(t0: number): string {
    let ds = trunc((getTime() - t0) / 1000);
    let minutes = trunc(ds / 60);
    let seconds = String(trunc(ds) % 60);
    while (seconds.length < 2) { seconds = '0' + seconds; }
    return '(' + minutes + ':' + seconds + ')';
}


interface Suite {
    title: string;
    file: string;
    suites: Array<Suite>;
    _t0: number;
    _countFail: number;
    _countPass: number;
    _countTotal: number;
    _countSkip: number;
}


interface Runner {
  on(event: string, callback: (...args: Array<any>) => void): Runner;
}


export type LogFunc = (message: string) => void;


let _logFunc: LogFunc = console.log.bind(console);
export function setLogFunc(logFunc: LogFunc) {
    _logFunc = logFunc
}


export function Reporter(runner: Runner) {


    let suites: Array<Suite> = [];


    // Force Output; Keeps the console output alive with periodic updates
    let lastOutput = getTime();
    function forceOutput() {
        if (((getTime() - lastOutput) / 1000) > MAX_DELAY) {
            const currentSuite = suites[suites.length - 1];
            log(`[ Still running suite - test # ${ (currentSuite ? currentSuite._countTotal: "0") } ]`);
        }
    }
    const timer = setInterval(forceOutput, 1000);




    function getIndent(): string {
        let result = '';
        for (let i = 0; i < suites.length; i++) { result += '  '; }
        return result;
    }


    function log(message?: string): void {
        if (!message) { message = ''; }
        _logFunc(getIndent() + message);
        lastOutput = getTime();
    }


    runner.on('suite', function(suite: Suite) {
        if (!suite.title) {
            log('Testing: ' + (suite.suites ? 'Found ' + suite.suites.length + ' test suites' : ''));
        } else {
            let filename = (suite.file || '').split('/').pop();
            if (filename) { filename = ' (' + filename + ')'; }
            log('Test Suite: ' + suite.title + filename);
        }


        suites.push(suite);


        suite._t0 = getTime();
        suite._countFail = 0;
        suite._countPass = 0;
        suite._countSkip = 0;
        suite._countTotal = 0;
    });


    runner.on('suite end', function() {
        let suite = suites.pop();


        let extras = [ ];


        if (suite._countSkip) {
            extras.push(suite._countSkip + " skipped");
        }


        if (suite._countTotal > suite._countPass) {
            extras.push((suite._countTotal - suite._countPass) + " failed");
        }


        let extra = "";
        if (extras.length) {
            extra = " (" + extras.join(",") + ")  ******** WARNING! ********";
        }


        log(`  Total Tests: ${ suite._countPass }/${ suite._countTotal } passed ${ getDelta(suite._t0) } ${ extra} \n`);


        if (suites.length > 0) {
            let currentSuite = suites[suites.length - 1];
            currentSuite._countFail += suite._countFail;
            currentSuite._countPass += suite._countPass;
            currentSuite._countSkip += suite._countSkip;
            currentSuite._countTotal += suite._countTotal;
        } else {
            clearTimeout(timer);
            const status = (suite._countPass === suite._countTotal) ? 0: 1;
            log(`# status:${ status }`);


            // Force quit after 5s
            setTimeout(() => {
                process.exit(status);
            }, 5000);
        }
    });


    runner.on('test', function(test) {
        forceOutput();
        if (test._currentRetry === 0) {
            const currentSuite = suites[suites.length - 1];
            currentSuite._countTotal++;
        }
    });


    runner.on('fail', function(test, error) {
        let currentSuite = suites[suites.length - 1];
        currentSuite._countFail++;


        let countFail = currentSuite._countFail;


        if (countFail > 100) {
            if (countFail === 101) {
                log('[ Over 100 errors; skipping remaining suite output ]');
            }
            return;
        }


        if (countFail > 25) {
            log('Failure: ' + test.title + ' (too many errors; skipping dump)');
            return;
        }


        log('Failure: ' + test.title);


        error.toString().split('\n').forEach((line: string) => {
            log('  ' + line);
        });


        Object.keys(error).forEach(function(key) {
            log('  ' + key + ': ' + error[key]);
        });


        if (error.stack) {
            log("Stack Trace:");
            error.stack.split('\n').forEach((line: string) => {
                log('  ' + line);
            });
        }
    });


    runner.on('pass', function(test) {
        let currentSuite = suites[suites.length - 1];
        currentSuite._countPass++;
    });


    runner.on('pending', function(test) {
        let currentSuite = suites[suites.length - 1];
        currentSuite._countSkip++;
    });
}


src/index.ts
import { BlockTag, TransactionReceipt, TransactionRequest, } from '@ethersproject/abstract-provider';
import { Networkish } from '@ethersproject/networks';
import { BaseProvider, TransactionResponse } from '@ethersproject/providers';
import { ConnectionInfo, fetchJson } from '@ethersproject/web';
import { BigNumber } from '@ethersproject/bignumber';
import { Signer } from '@ethersproject/abstract-signer';
import * as providers from '@ethersproject/providers';
import { encode } from '@ethersproject/rlp';
import { encrypt } from 'eciesjs';
import { _TypedDataEncoder, id } from '@ethersproject/hash';
import { keccak256 } from '@ethersproject/keccak256';
import { parse as parseTransaction } from '@ethersproject/transactions';
import { parseUnits } from '@ethersproject/units';

export const DEFAULT_FLASHBOTS_RELAY = 'https://relay.flashbots.net';
export const BASE_FEE_MAX_CHANGE_DENOMINATOR = 8;

export enum FlashbotsBundleResolution {
  BundleIncluded,
  BlockPassedWithoutInclusion,
  AccountNonceTooHigh,
}

export interface FlashbotsBundleRawTransaction {
  signedTransaction: string;
}

export interface FlashbotsBundleTransaction {
  transaction: TransactionRequest;
  signer: Signer;
}

export interface FlashbotsOptions {
  minTimestamp?: number;
  maxTimestamp?: number;
  revertingTxHashes?: Array<string>;
}

export interface FlashbotsBundle {
  signedBundledTransactions: Array<string>;
  blockTarget: number;
  options?: FlashbotsOptions;
}

export interface TransactionAccountNonce {
  hash: string;
  signedTransaction: string;
  account: string;
  nonce: number;
}

export interface FlashbotsTransactionResponse {
  bundleTransactions: Array<TransactionAccountNonce>;
  wait: () => Promise<FlashbotsBundleResolution>;
  simulate: () => Promise<SimulationResponse>;
  receipts: () => Promise<Array<TransactionReceipt>>;
}

export interface TransactionSimulationBase {
  txHash: string;
  gasUsed: number;
}

export interface TransactionSimulationSuccess
  extends TransactionSimulationBase {
  value: string;
}

export interface TransactionSimulationRevert extends TransactionSimulationBase {
  error: string;
  revert: string;
}

export type TransactionSimulation =
  | TransactionSimulationSuccess
  | TransactionSimulationRevert;

export interface RelayResponseError {
  error: {
    message: string;
    code: number;
  };
}

export interface SimulationResponseSuccess {
  bundleHash: string;
  coinbaseDiff: BigNumber;
  results: Array<TransactionSimulation>;
  totalGasUsed: number;
  firstRevert?: TransactionSimulation;
}

export type SimulationResponse = SimulationResponseSuccess | RelayResponseError;

export type FlashbotsTransaction =
  | FlashbotsTransactionResponse
  | RelayResponseError;

export interface GetUserStatsResponseSuccess {
  signing_address: string;
  blocks_won_total: number;
  bundles_submitted_total: number;
  bundles_error_total: number;
  avg_gas_price_gwei: number;
  blocks_won_last_7d: number;
  bundles_submitted_last_7d: number;
  bundles_error_7d: number;
  avg_gas_price_gwei_last_7d: number;
  blocks_won_last_numberd: number;
  bundles_submitted_last_numberd: number;
  bundles_error_numberd: number;
  avg_gas_price_gwei_last_numberd: number;
  blocks_won_last_numberh: number;
  bundles_submitted_last_numberh: number;
  bundles_error_numberh: number;
  avg_gas_price_gwei_last_numberh: number;
  blocks_won_last_5m: number;
  bundles_submitted_last_5m: number;
  bundles_error_5m: number;
  avg_gas_price_gwei_last_5m: number;
}

export type GetUserStatsResponse =
  | GetUserStatsResponseSuccess
  | RelayResponseError;

export interface GetBundleStatsResponseSuccess {
  isSimulated: boolean;
  isSentToMiners: boolean;
  isHighPriority: boolean;
  simulatedAt: string;
  submittedAt: string;
  sentToMinersAt: string;
}

export type GetBundleStatsResponse =
  | GetBundleStatsResponseSuccess
  | RelayResponseError;

type RpcParams = Array<string[] | string | number | Record<string, unknown>>;

const TIMEOUT_MS = 5 * 60 * 1000;

export class FlashbotsBundleProvider extends providers.JsonRpcProvider {
  private genericProvider: BaseProvider;
  private authSigner: Signer;
  private connectionInfo: ConnectionInfo;

  constructor(
    genericProvider: BaseProvider,
    authSigner: Signer,
    connectionInfoOrUrl: ConnectionInfo,
    network: Networkish,
  ) {
    super(connectionInfoOrUrl, network);
    this.genericProvider = genericProvider;
    this.authSigner = authSigner;
    this.connectionInfo = connectionInfoOrUrl;
  }

  static async throttleCallback(): Promise<boolean> {
    console.warn('Rate limited');
    return false;
  }

  static async create(
    genericProvider: BaseProvider,
    authSigner: Signer,
    connectionInfoOrUrl?: ConnectionInfo | string,
    network?: Networkish,
  ): Promise<FlashbotsBundleProvider> {
    const connectionInfo: ConnectionInfo =
      typeof connectionInfoOrUrl === 'string' ||
      typeof connectionInfoOrUrl === 'undefined'
        ? {
            url: connectionInfoOrUrl || DEFAULT_FLASHBOTS_RELAY,
          }
        : {
            ...connectionInfoOrUrl,
          };
    if (connectionInfo.headers === undefined) connectionInfo.headers = {};
    connectionInfo.throttleCallback = FlashbotsBundleProvider.throttleCallback;
    const networkish: Networkish = {
      chainId: 0,
      name: '',
    };
    if (typeof network === 'string') {
      networkish.name = network;
    } else if (typeof network === 'number') {
      networkish.chainId = network;
    } else if (typeof network === 'object') {
      networkish.name = network.name;
      networkish.chainId = network.chainId;
    }

    if (networkish.chainId === 0) {
      networkish.chainId = (await genericProvider.getNetwork()).chainId;
    }

    return new FlashbotsBundleProvider(
      genericProvider,
      authSigner,
      connectionInfo,
      networkish,
    );
  }

  static getMaxBaseFeeInFutureBlock(
    baseFee: BigNumber,
    blocksInFuture: number,
  ): BigNumber {
    let maxBaseFee = BigNumber.from(baseFee);
    for (let i = 0; i < blocksInFuture; i++) {
      maxBaseFee = maxBaseFee.mul(1125).div(1000).add(1);
    }
    return maxBaseFee;
  }

  static getBaseFeeInNextBlock(
    currentBaseFeePerGas: BigNumber,
    currentGasUsed: BigNumber,
    currentGasLimit: BigNumber,
  ): BigNumber {
    const currentGasTarget = currentGasLimit.div(2);

    if (currentGasUsed.eq(currentGasTarget)) {
      return currentBaseFeePerGas;
    } else if (currentGasUsed.gt(currentGasTarget)) {
      const gasUsedDelta = currentGasUsed.sub(currentGasTarget);
      const baseFeePerGasDelta = currentBaseFeePerGas
        .mul(gasUsedDelta)
        .div(currentGasTarget)
        .div(BASE_FEE_MAX_CHANGE_DENOMINATOR);

      return currentBaseFeePerGas.add(baseFeePerGasDelta);
    } else {
      const gasUsedDelta = currentGasTarget.sub(currentGasUsed);
      const baseFeePerGasDelta = currentBaseFeePerGas
        .mul(gasUsedDelta)
        .div(currentGasTarget)
        .div(BASE_FEE_MAX_CHANGE_DENOMINATOR);

      return currentBaseFeePerGas.sub(baseFeePerGasDelta);
    }
  }

  static generateBundleHash(txHashes: Array<string>): string {
    const concatenatedHashes = txHashes
      .map((txHash) => txHash.slice(2))
      .join('');
    return keccak256(`0x${concatenatedHashes}`);
  }
  public async sendRawBundle(
    signedBundledTransactions: Array<string>,
    targetBlockNumber: number,
    opts?: FlashbotsOptions,
  ): Promise<FlashbotsTransaction> {
    const params = {
      txs: signedBundledTransactions,
      blockNumber: `0x${targetBlockNumber.toString(16)}`,
      minTimestamp: opts?.minTimestamp,
      maxTimestamp: opts?.maxTimestamp,
      revertingTxHashes: opts?.revertingTxHashes,
    };

    const request = JSON.stringify(
      this.prepareBundleRequest('eth_sendBundle', [params]),
    );
    const response = await this.request(request);
    if (response.error !== undefined && response.error !== null) {
      return {
        error: {
          message: response.error.message,
          code: response.error.code,
        },
      };
    }

    const bundleTransactions = signedBundledTransactions.map(
      (signedTransaction) => {
        const transactionDetails = parseTransaction(signedTransaction);
        return {
          signedTransaction,
          hash: keccak256(signedTransaction),
          account: transactionDetails.from || '0x0',
          nonce: transactionDetails.nonce,
        };
      },
    );

    return {
      bundleTransactions,
      wait: () => this.wait(bundleTransactions, targetBlockNumber, TIMEOUT_MS),
      simulate: () =>
        this.simulate(
          bundleTransactions.map((tx) => tx.signedTransaction),
          targetBlockNumber,
          undefined,
          opts?.minTimestamp,
        ),
      receipts: () => this.fetchReceipts(bundleTransactions),
    };
  }

  public async sendBundle(
    bundledTransactions: Array<
      FlashbotsBundleTransaction | FlashbotsBundleRawTransaction
    >,
    targetBlockNumber: number,
    opts?: FlashbotsOptions,
  ): Promise<FlashbotsTransaction> {
    const signedTransactions = await this.signBundle(bundledTransactions);
    return this.sendRawBundle(signedTransactions, targetBlockNumber, opts);
  }

  public async signBundle(
    bundledTransactions: Array<
      FlashbotsBundleTransaction | FlashbotsBundleRawTransaction
    >,
  ): Promise<Array<string>> {
    const nonces: { [address: string]: BigNumber } = {};
    const signedTransactions = new Array<string>();
    for (const tx of bundledTransactions) {
      if ('signedTransaction' in tx) {
        // in case someone is mixing pre-signed and signing transactions, decode to add to nonce object
        const transactionDetails = parseTransaction(tx.signedTransaction);
        if (transactionDetails.from === undefined)
          throw new Error('Could not decode signed transaction');
        nonces[transactionDetails.from] = BigNumber.from(
          transactionDetails.nonce + 1,
        );
        signedTransactions.push(tx.signedTransaction);
        continue;
      }
      const transaction = { ...tx.transaction };
      const address = await tx.signer.getAddress();
      if (typeof transaction.nonce === 'string') throw new Error('Bad nonce');
      const nonce =
        transaction.nonce !== undefined
          ? BigNumber.from(transaction.nonce)
          : nonces[address] ||
            BigNumber.from(
              await this.genericProvider.getTransactionCount(address, 'latest'),
            );
      nonces[address] = nonce.add(1);
      if (transaction.nonce === undefined) transaction.nonce = nonce;
      if (
        (transaction.type == null || transaction.type == 0) &&
        transaction.gasPrice === undefined
      )
        transaction.gasPrice = BigNumber.from(0);
      if (transaction.gasLimit === undefined)
        transaction.gasLimit = await tx.signer.estimateGas(transaction); // TODO: Add target block number and timestamp when supported by geth
      signedTransactions.push(await tx.signer.signTransaction(transaction));
    }
    return signedTransactions;
  }

  private wait(
    transactionAccountNonces: Array<TransactionAccountNonce>,
    targetBlockNumber: number,
    timeout: number,
  ) {
    return new Promise<FlashbotsBundleResolution>((resolve, reject) => {
      let timer: NodeJS.Timer | null = null;
      let done = false;

      const minimumNonceByAccount = transactionAccountNonces.reduce(
        (acc, accountNonce) => {
          if (
            accountNonce.nonce > 0 &&
            (accountNonce.nonce || 0) < acc[accountNonce.account]
          ) {
            acc[accountNonce.account] = accountNonce.nonce;
          }
          acc[accountNonce.account] = accountNonce.nonce;
          return acc;
        },
        {} as { [account: string]: number },
      );
      const handler = async (blockNumber: number) => {
        if (blockNumber < targetBlockNumber) {
          const noncesValid = await Promise.all(
            Object.entries(minimumNonceByAccount).map(
              async ([account, nonce]) => {
                const transactionCount =
                  await this.genericProvider.getTransactionCount(account);
                return nonce >= transactionCount;
              },
            ),
          );
          const allNoncesValid = noncesValid.every(Boolean);
          if (allNoncesValid) return;
          // target block not yet reached, but nonce has become invalid
          resolve(FlashbotsBundleResolution.AccountNonceTooHigh);
        } else {
          const block = await this.genericProvider.getBlock(targetBlockNumber);
          // check bundle against block:
          const blockTransactionsHash: { [key: string]: boolean } = {};
          for (const bt of block.transactions) {
            blockTransactionsHash[bt] = true;
          }
          const bundleIncluded = transactionAccountNonces.every(
            (transaction) => blockTransactionsHash[transaction.hash],
          );
          resolve(
            bundleIncluded
              ? FlashbotsBundleResolution.BundleIncluded
              : FlashbotsBundleResolution.BlockPassedWithoutInclusion,
          );
        }

        if (timer) {
          clearTimeout(timer);
        }
        if (done) {
          return;
        }
        done = true;

        this.genericProvider.removeListener('block', handler);
      };
      this.genericProvider.on('block', handler);

      if (typeof timeout === 'number' && timeout > 0) {
        timer = setTimeout(() => {
          if (done) {
            return;
          }
          timer = null;
          done = true;

          this.genericProvider.removeListener('block', handler);
          reject('Timed out');
        }, timeout);
        if (timer.unref) {
          timer.unref();
        }
      }
    });
  }

  public async getUserStats(): Promise<GetUserStatsResponse> {
    const blockDetails = await this.genericProvider.getBlock('latest');
    const evmBlockNumber = `0x${blockDetails.number.toString(16)}`;

    const params = [evmBlockNumber];
    const request = JSON.stringify(
      this.prepareBundleRequest('flashbots_getUserStats', params),
    );
    const response = await this.request(request);
    if (response.error !== undefined && response.error !== null) {
      return {
        error: {
          message: response.error.message,
          code: response.error.code,
        },
      };
    }

    return response.result;
  }

  public async getBundleStats(
    bundleHash: string,
    blockNumber: number,
  ): Promise<GetBundleStatsResponse> {
    const evmBlockNumber = `0x${blockNumber.toString(16)}`;

    const params = [{ bundleHash, blockNumber: evmBlockNumber }];
    const request = JSON.stringify(
      this.prepareBundleRequest('flashbots_getBundleStats', params),
    );
    const response = await this.request(request);
    if (response.error !== undefined && response.error !== null) {
      return {
        error: {
          message: response.error.message,
          code: response.error.code,
        },
      };
    }

    return response.result;
  }

  public async simulate(
    signedBundledTransactions: Array<string>,
    blockTag: BlockTag,
    stateBlockTag?: BlockTag,
    blockTimestamp?: number,
  ): Promise<SimulationResponse> {
    let evmBlockNumber: string;
    if (typeof blockTag === 'number') {
      evmBlockNumber = `0x${blockTag.toString(16)}`;
    } else {
      const blockTagDetails = await this.genericProvider.getBlock(blockTag);
      const blockDetails =
        blockTagDetails !== null
          ? blockTagDetails
          : await this.genericProvider.getBlock('latest');
      evmBlockNumber = `0x${blockDetails.number.toString(16)}`;
    }

    let evmBlockStateNumber: string;
    if (typeof stateBlockTag === 'number') {
      evmBlockStateNumber = `0x${stateBlockTag.toString(16)}`;
    } else if (!stateBlockTag) {
      evmBlockStateNumber = 'latest';
    } else {
      evmBlockStateNumber = stateBlockTag;
    }

    const params: RpcParams = [
      {
        txs: signedBundledTransactions,
        blockNumber: evmBlockNumber,
        stateBlockNumber: evmBlockStateNumber,
        timestamp: blockTimestamp,
      },
    ];
    const request = JSON.stringify(
      this.prepareBundleRequest('eth_callBundle', params),
    );
    const response = await this.request(request);
    if (response.error !== undefined && response.error !== null) {
      return {
        error: {
          message: response.error.message,
          code: response.error.code,
        },
      };
    }

    const callResult = response.result;
    return {
      bundleHash: callResult.bundleHash,
      coinbaseDiff: BigNumber.from(callResult.coinbaseDiff),
      results: callResult.results,
      totalGasUsed: callResult.results.reduce(
        (a: number, b: TransactionSimulation) => a + b.gasUsed,
        0,
      ),
      firstRevert: callResult.results.find(
        (txSim: TransactionSimulation) => 'revert' in txSim,
      ),
    };
  }

  /**
   * Method to send a carrier tx into the public mempool
   *
   * @param bundle  FlashbotsBundle with AT LEAST signed bundled transactions in signedBundledTransactions field obtained
   *  from {@link signBundle} method, and blockTarget.
   * @param validatorPublicKey  The public key of the validator that will be able to decrypt the bundle and include it
   *  into the bundle pool.
   * @param signer  Signer who will sign the carrier transaction.
   * @param carrierTx TransactionRequest whose data field will carry the encrypted bundle : MAY be an incomplete
   *  object which will be populated with default values.
   *
   * @return Promise<TransactionResponse> Promise containing the response for the carrier tx
   * */

  public async sendCarrierTransaction(
    bundle: FlashbotsBundle,
    validatorPublicKey: string,
    signer: Signer,
    carrierTx: TransactionRequest,
  ): Promise<TransactionResponse> {
    //RLP-serialize the given bundle
    const serializedBundle = this.rlpSerializeBundle(bundle);

    //Encrypt the encoded bundle with the passed validator pub_key
    const encryptedBundle = encrypt(
      validatorPublicKey,
      Buffer.from(serializedBundle),
    );

    //Populate carrier_tx.data as : carrier_tx.data = MEV_Prefix | validator pub_key | Encrypt(validator pub_key, serialized bundle)
    const mevPrefix = `0123`; //this is a placeholder!

    let payload = `0x`;
    payload += mevPrefix;
    payload += validatorPublicKey;
    payload += encryptedBundle.toString('hex');

    carrierTx.data = payload;

    //Check if carrier_tx has minimum params, populate with defaults if not
    /*
     The following statement is intended to be used in order to support any type of incomplete TransactionRequest
     received, populating it with default values if any one is missing
     */
    await this.populateCarrierTransaction(carrierTx, signer);

    //Sign the transaction received as param with passed signer
    const signedTx = await signer.signTransaction(carrierTx);

    //Propagate carrier_tx into the public mempool and return Promise<TransactionResponse> for the carrier_tx
    return this.genericProvider.sendTransaction(signedTx);
  }

  /**
   * A private method to encode a FlashbotsBundle following the RLP serialization standard
   * @param bundle the FlashbotsBundle instance to be serialized
   * @return string the rlp encoded bundle
   * @private
   */
  private rlpSerializeBundle(bundle: FlashbotsBundle): string {
    if (
      bundle.signedBundledTransactions === undefined ||
      bundle.signedBundledTransactions.length === 0
    )
      throw Error('Bundle has no transactions');
    if (bundle.options === undefined) bundle.options = {};

    const fields = [
      bundle.signedBundledTransactions,
      this.formatNumber(bundle.blockTarget || 0),
      this.formatNumber(bundle.options.minTimestamp || 0),
      this.formatNumber(bundle.options.maxTimestamp || 0),
      bundle.options.revertingTxHashes || [],
    ];
    return encode(fields);
  }

  private formatNumber(num: number): string {
    const hexNum = num.toString(16);
    return hexNum.length % 2 === 0 ? `0x${hexNum}` : `0x0${hexNum}`;
  }

  /**
   * A private method to populate {@param carrier}'s missing fields with default values
   * @param carrier an instance of TransactionRequest which will be the tx containing the full payload in its data field
   * @param signer the signer Object which will send the carrier tx
   * @private
   */
  private async populateCarrierTransaction(
    carrier: TransactionRequest,
    signer: Signer,
  ) {
    if (!('to' in carrier)) throw Error('carrier.to field is missing');

    if (carrier.gasPrice != null) {
      const gasPrice = BigNumber.from(carrier.gasPrice);
      const maxFeePerGas = BigNumber.from(carrier.maxFeePerGas || 0);
      if (!gasPrice.eq(maxFeePerGas)) {
        throw Error('carrier tx EIP-1559 mismatch: gasPrice != maxFeePerGas');
      }
    }
    const latestBlock = await this.genericProvider.getBlock('latest');
    const blocksInFuture = 5;
    const maxBaseFeeInFuture =
      FlashbotsBundleProvider.getMaxBaseFeeInFutureBlock(
        <BigNumber>latestBlock.baseFeePerGas,
        blocksInFuture,
      );

    carrier.type = 2;
    carrier.chainId = carrier.chainId || 1;
    carrier.nonce =
      carrier.nonce ||
      (await this.genericProvider.getTransactionCount(signer.getAddress()));
    carrier.maxPriorityFeePerGas =
      carrier.maxPriorityFeePerGas || parseUnits('1.5', 'gwei');
    carrier.maxFeePerGas =
      carrier.maxFeePerGas ||
      maxBaseFeeInFuture.add(carrier.maxPriorityFeePerGas);
    carrier.gasLimit =
      carrier.gasLimit || (await this.genericProvider.estimateGas(carrier));
    carrier.value = carrier.value || 0;
    carrier.accessList = carrier.accessList || [];
  }

  private async request(request: string) {
    const connectionInfo = { ...this.connectionInfo };
    connectionInfo.headers = {
      'X-Flashbots-Signature': `${await this.authSigner.getAddress()}:${await this.authSigner.signMessage(
        id(request),
      )}`,
      ...this.connectionInfo.headers,
    };
    return fetchJson(connectionInfo, request);
  }

  private async fetchReceipts(
    bundledTransactions: Array<TransactionAccountNonce>,
  ): Promise<Array<TransactionReceipt>> {
    return Promise.all(
      bundledTransactions.map((bundledTransaction) =>
        this.genericProvider.getTransactionReceipt(bundledTransaction.hash),
      ),
    );
  }

  private prepareBundleRequest(
    method:
      | 'eth_callBundle'
      | 'eth_sendBundle'
      | 'flashbots_getUserStats'
      | 'flashbots_getBundleStats',
    params: RpcParams,
  ) {
    return {
      method: method,
      params: params,
      id: this._nextId++,
      jsonrpc: '2.0',
    };
  }
}
